VDA-StorageNode-Registry/contracts/StorageNodeRegistry.sol

```diff
@@ -0,0 +1,1098 @@
+     * @param didAddress DID address
+     * @return uint Return 0 if staked amount is less than the required amount
+     */
+    function getExcessTokenAmount(address didAddress) internal view returns(uint) {
+        uint totalAmount;
+
+        uint nodeId = _didNodeId[didAddress];
+        if (nodeId != 0) {
+            totalAmount = requiredTokenAmount(_nodeMap[nodeId].slotCount);    
+        }
+
+        if (_stakedTokenAmount[didAddress] <= totalAmount) {
+            return 0;
+        }
+
+        return (_stakedTokenAmount[didAddress] - totalAmount);
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function excessTokenAmount(address didAddress) external view returns(uint) {
+        return getExcessTokenAmount(didAddress);
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function withdraw(
+        address didAddress,
+        uint amount,
+        bytes calldata requestSignature,
+        bytes calldata requestProof
+    ) external {
+        {
+            bytes memory params = abi.encodePacked(didAddress, amount);
+            verifyRequest(didAddress, params, requestSignature, requestProof);
+        }
+
+        uint excessAmount = getExcessTokenAmount(didAddress);
+
+        if (excessAmount == 0) {
+            revert NoExcessTokenAmount();
+        }
+
+        if (amount > excessAmount) {
+            revert InvalidAmount();
+        }
+
+        IERC20Upgradeable(vdaTokenAddress).transfer(tx.origin, amount);
+
+        _stakedTokenAmount[didAddress] = _stakedTokenAmount[didAddress] - amount;
+
+        emit TokenWithdrawn(didAddress, tx.origin, amount);
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function depositToken(address didAddress, uint tokenAmount) external {
+        uint nodeId = _didNodeId[didAddress];
+        if (nodeId == 0) {
+            revert InvalidDIDAddress();
+        }
+
+        IERC20Upgradeable(vdaTokenAddress).transferFrom(tx.origin, address(this), tokenAmount);
+
+        _stakedTokenAmount[didAddress] = _stakedTokenAmount[didAddress] + tokenAmount;
+
+        emit TokenDeposited(didAddress, tx.origin, tokenAmount);
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function getNodeIssueFee() external view returns(uint){
+        return _slotInfo.NODE_ISSUE_FEE;
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function updateNodeIssueFee(uint value) external payable onlyOwner {
+        if (value == 0 || value == _slotInfo.NODE_ISSUE_FEE) {
+            revert InvalidValue();
+        }
+        uint orgFee = _slotInfo.NODE_ISSUE_FEE;
+        _slotInfo.NODE_ISSUE_FEE = value;
+
+        emit UpdateNodeIssueFee(orgFee, value);
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function getTotalIssueFee() external view returns(uint) {
+        return _slotInfo.totalIssueFee;
+    }
+
+    /**
```

Comment:
Add comment

Reply:
The comments are in the `IStorageNodeRegistry.sol` file.